File: 200.岛屿数量.go
/*
 * @lc app=leetcode.cn id=200 lang=golang
 *
 * [200] 岛屿数量
 */

// @lc code=start
var dir = [4][2]int{
	{-1, 0},
	{0, 1},
	{1, 0},
	{0, -1},
}

func numIslands(grid [][]byte) int {
	m := len(grid)
	if m == 0 {
		return 0
	}

	n := len(grid[0])
	if n == 0 {
		return 0
	}

	res, visited := 0, make([][]bool, m)
	for i := 0; i < m; i++ {
		visited[i] = make([]bool, n)
	}

	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if grid[i][j] == '1' && !visited[i][j] {
				searchIslands(grid, &visited, i, j)
				res++
			}
		}
	}
	return res
}

// 从一个陆地开始向四周搜索将陆地标记为 visited。
func searchIslands(grid [][]byte, visited *[][]bool, x, y int) {
	(*visited)[x][y] = true
	for i := 0; i < 4; i++ {
		nx := x + dir[i][0]
		ny := y + dir[i][1]
		if isInBoard(grid, nx, ny) && !(*visited)[nx][ny] && grid[nx][ny] == '1' {
			searchIslands(grid, visited, nx, ny)
		}
	}
}

func isInBoard(grid [][]byte, x, y int) bool {
	m := len(grid)
	if x >= m {
		return false
	}

	n := len(grid[0])
	if y >= n {
		return false
	}

	if x >= 0 && x < m && y >= 0 && y < n {
		return true
	}
	return false
}

// @lc code=end

File: 200.岛屿数量.java
/*
 * @lc app=leetcode.cn id=200 lang=java
 *
 * [200] 岛屿数量
 */

// @lc code=start
class Solution {
    private int[][] directions = new int[][] {
        {-1, 0},
        {0, -1},
        {1, 0},
        {0, 1},
    };

    public int numIslands(char[][] grid) {
        if (grid == null || grid.length == 0) {
            return 0;
        }

        int rows = grid.length;
        int cols = grid[0].length;
        int res = 0;
        int[][] visited = new int[rows][cols];
        for (int y = 0; y < rows; y++) {
            for (int x = 0; x < cols; x++) {
                if (grid[y][x] == '1' && visited[y][x] == 0) {
                    searchIslands(grid, visited, x, y);
                    res++;
                }
            }
        }
        return res;
    }

    // Note: 连成一片的算一个岛。
    private void searchIslands(char[][] grid, int[][] visited, int x, int y) {
        visited[y][x] = 1;
        for (int i = 0; i < 4; i++) {
            int nx = x + directions[i][0];
            int ny = y + directions[i][1];
            if (inBoard(grid, nx, ny) && visited[ny][nx] == 0 && grid[ny][nx] == '1') {
                searchIslands(grid, visited, nx, ny);
            }
        }
    }

    private boolean inBoard(char[][] grid, int x, int y) {
        int rows = grid.length;
        int cols = grid[0].length;
        return 0 <= x && x < cols && 0 <= y && y < rows;
    }
}
// @lc code=end

